File: src/dev_health_ops/api/graphql/resolvers/review_edges.py
"""Review Edges GraphQL resolver (CHAOS-2077).

Reads from the append-only ClickHouse table ``review_edges_daily``.
Returns reviewer-to-author collaboration edges filtered by org, date range,
and optionally by repo.  All reads are org-scoped via ``require_org_id``.
No data is written or recomputed.
"""

from __future__ import annotations

import logging
from typing import Any

from dev_health_ops.api.queries.client import query_dicts

from ..authz import require_org_id
from ..context import GraphQLContext
from ..types.review_edges import (
    ReviewEdgeRow,
    ReviewEdgesInput,
    ReviewEdgesResult,
)

logger = logging.getLogger(__name__)

#: Hard cap on returned rows to protect against pathological date ranges.
MAX_REVIEW_EDGES_ROWS: int = 2000


def _require_client(context: GraphQLContext) -> Any:
    if context.client is None:
        raise RuntimeError("Database client not available for ReviewEdges resolver")
    return context.client


# ---------------------------------------------------------------------------
# ClickHouse fetch helper
# ---------------------------------------------------------------------------


async def _fetch_review_edges(
    client: Any,
    *,
    org_id: str,
    since_date: str,
    until_date: str,
    repo_ids: list[str] | None,
    limit: int,
) -> list[dict[str, Any]]:
    """Read ``review_edges_daily`` filtered by org + date range.

    ``review_edges_daily`` is append-only (plain MergeTree, NOT
    ReplacingMergeTree): a recompute / backfill writes a duplicate row for the
    same ``(org_id, repo_id, reviewer, author, day)`` key (live data shows 49
    duplicate-key rows). The inner subquery collapses each key to its latest
    row via ``argMax(reviews_count, computed_at)`` so a backfilled day is not
    counted twice. Only after deduplication do we ``ORDER BY reviews_count
    DESC`` (heaviest collaboration pairs first) and apply the row cap.

    An optional ``repo_ids`` filter narrows to specific repositories.
    ``repo_id`` is stored as a UUID; input repo refs may be UUID strings or
    ``repos.repo`` slugs, so the predicate resolves them through the org-scoped
    catalog before comparing UUID-to-UUID.

    CHAOS-4421 / CHAOS-4368 Part A: ``ORDER BY reviews_count DESC`` alone has
    no tie-breaker. ClickHouse does NOT guarantee a stable row order among
    rows with equal ``reviews_count`` -- an ``ORDER BY`` with ties, combined
    with a ``LIMIT`` sitting at a tie boundary, can return a different row
    set/order across otherwise-identical runs (parallel block processing has
    no ordering guarantee beyond the declared sort key; see ClickHouse docs
    on ORDER BY: "If ... rows have the same value ... the resulting order of
    such rows is undefined and may be non-deterministic"). That makes this
    resolver's output genuinely non-deterministic at a tie boundary, which
    the Go/Python parity comparator (CHAOS-4381 stage 2) cannot tolerate --
    comparing two non-deterministic outputs is not a valid proof of parity.
    Fix: extend ``ORDER BY`` with a deterministic tie-break drawn from the
    row's own key columns -- ``repo_id, reviewer, author, day`` is exactly
    the ``GROUP BY`` key above, so it is already a total order over the
    deduplicated row set; no synthetic tiebreaker column is introduced.
    """
    inner_where = """
            WHERE org_id = {org_id:String}
              AND day >= {since_date:Date}
              AND day <= {until_date:Date}
    """
    params: dict[str, Any] = {
        "org_id": org_id,
        "since_date": since_date,
        "until_date": until_date,
    }
    if repo_ids:
        inner_where += """
              AND repo_id IN (
                  SELECT id FROM repos
                  WHERE org_id = {org_id:String}
                    AND (repo IN {repo_ids:Array(String)} OR toString(id) IN {repo_ids:Array(String)})
              )"""
        params["repo_ids"] = list(repo_ids)

    query = f"""
        SELECT
            reviewer,
            author,
            reviews_count,
            day,
            toString(repo_id) AS repo_id
        FROM (
            SELECT
                repo_id,
                reviewer,
                author,
                day,
                argMax(reviews_count, computed_at) AS reviews_count
            FROM review_edges_daily
            {inner_where}
            GROUP BY repo_id, reviewer, author, day
        )
        ORDER BY reviews_count DESC, repo_id, reviewer, author, day
        LIMIT {limit}
    """
    return await query_dicts(client, query, params)


# ---------------------------------------------------------------------------
# Public resolver
# ---------------------------------------------------------------------------


async def resolve_review_edges(
    context: GraphQLContext,
    input: ReviewEdgesInput,
) -> ReviewEdgesResult:
    """Serve review-edge rows from ClickHouse (read-only).

    Org-gate is enforced via ``require_org_id``; any mismatch between the
    JWT org and the GraphQL ``orgId`` argument is logged and the JWT org wins.
    """
    authorized_org_id = require_org_id(context)
    if input.org_id != authorized_org_id:
        logger.debug(
            "Ignoring GraphQL orgId %r in favor of authorized org %r",
            input.org_id,
            authorized_org_id,
        )

    client = _require_client(context)

    raw_limit = input.limit if input.limit is not None else 500
    effective_limit = max(1, min(raw_limit, MAX_REVIEW_EDGES_ROWS))

    since_date = input.since_date.isoformat()
    until_date = input.until_date.isoformat()

    raw_rows = await _fetch_review_edges(
        client,
        org_id=authorized_org_id,
        since_date=since_date,
        until_date=until_date,
        repo_ids=input.repo_ids,
        limit=effective_limit,
    )

    edges: list[ReviewEdgeRow] = []
    for row in raw_rows:
        edges.append(
            ReviewEdgeRow(
                reviewer=str(row.get("reviewer") or ""),
                author=str(row.get("author") or ""),
                reviews_count=int(row.get("reviews_count") or 0),
                day=row["day"],
                repo_id=str(row["repo_id"]) if row.get("repo_id") else None,
            )
        )

    return ReviewEdgesResult(edges=edges, total_count=len(edges))

File: cmd/query-api/internal/reviewedges/reviewedges.go
// Package reviewedges is the Go port of
// dev_health_ops.api.graphql.resolvers.review_edges.resolve_review_edges
// (ops/src/dev_health_ops/api/graphql/resolvers/review_edges.py),
// CHAOS-4352 plan §6 Wave 2's canary operation, ported after CHAOS-4368
// Part A (#1980, commit 8d34d8b6e) made the Python resolver's row order
// deterministic -- a prerequisite for the Go/Python parity comparator,
// which cannot validly compare two non-deterministic outputs.
//
// Ported deliberately verbatim: same inner-subquery dedup
// (`argMax(reviews_count, computed_at)` per `(repo_id, reviewer, author,
// day)`, since `review_edges_daily` is append-only plain MergeTree, not
// ReplacingMergeTree -- a recompute/backfill can duplicate a key), same
// `ORDER BY reviews_count DESC, repo_id, reviewer, author, day` (the
// Part-A deterministic tie-break -- the resolver's own GROUP BY key,
// already a total order over the deduplicated row set), same optional
// repo_ids filter (resolved through the org-scoped `repos` catalog by
// slug OR UUID string, exactly as `_fetch_review_edges` does), and the
// same limit clamp (1..2000, `MAX_REVIEW_EDGES_ROWS`).
//
// Side effects: none to replicate. Verified by reading
// `resolve_review_edges`/`_fetch_review_edges` top to bottom: one
// read-only ClickHouse query via `query_dicts` and a dataclass
// construction -- no telemetry/audit hook call inside it or anything it
// calls (same finding class as `featureflags`'s doc comment; unlike
// `home`/investment analytics, which plan §5 calls out by name).
//
// Missing-table behavior deliberately differs from `featureflags`: unlike
// `resolve_feature_flags`, `resolve_review_edges` has NO try/except around
// its ClickHouse call and `ReviewEdgesResult` has no `degradedReason`
// field at all -- a missing `review_edges_daily` table is a real error on
// the Python side, not a degraded empty result. This port does not invent
// a degraded path Python doesn't have; a ClickHouse error propagates as a
// Go error (and, via schema.resolvers.go, a GraphQL error), matching
// Python's actual behavior exactly.
package reviewedges

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/full-chaos/dev-health-go/clickhouse"

	"github.com/full-chaos/dev-health-ops/cmd/query-api/internal/graph/model"
	"github.com/full-chaos/dev-health-ops/cmd/query-api/internal/graphqldate"
)

// MaxRows mirrors Python's MAX_REVIEW_EDGES_ROWS -- a hard cap on returned
// rows to protect against pathological date ranges.
const MaxRows = 2000

// QueryClient is the read-only ClickHouse query boundary this package
// needs -- same shape as featureflags.QueryClient (declared independently
// per that package's own doc comment: this operation's query shape does
// not fit dev-health-go/readers' id-keyed convention, and adding a reader
// there is a separate, version-bump-owning change owned by the worker
// orchestrator, not this lane). *clickhouse.Client satisfies this
// interface directly.
type QueryClient interface {
	Query(ctx context.Context, statement string, bindings []clickhouse.Binding) (clickhouse.RowScanner, error)
}

// clampLimit bounds a caller-supplied limit to a safe 1..MaxRows range,
// mirroring resolve_review_edges's `effective_limit = max(1, min(raw_limit,
// MAX_REVIEW_EDGES_ROWS))`.
func clampLimit(limit int) int {
	if limit < 1 {
		return 1
	}
	if limit > MaxRows {
		return MaxRows
	}
	return limit
}

// Resolve ports resolve_review_edges/_fetch_review_edges. orgID must
// already be the AUTHORIZED org (the caller's verified envelope claim,
// not necessarily the client-supplied `input.orgId` GraphQL argument --
// see schema.resolvers.go's ReviewEdges for why: Python's resolver
// silently prefers the authorized org over a mismatched GraphQL argument
// rather than erroring, and this port reproduces that same
// "authorized org always wins" behavior by construction, taking only one
// org parameter). limit is clamped internally, same as Python -- callers
// must not pre-clamp and must not trust the GraphQL schema's default
// alone (a client can send any value).
func Resolve(ctx context.Context, client QueryClient, orgID string, sinceDate, untilDate graphqldate.Date, repoIDs []string, limit int) (*model.ReviewEdgesResult, error) {
	if client == nil {
		return nil, errors.New("reviewedges: clickhouse client is required")
	}

	bindings := []clickhouse.Binding{
		{Name: "org_id", Value: orgID},
		{Name: "since_date", Value: sinceDate.String()},
		{Name: "until_date", Value: untilDate.String()},
		{Name: "limit", Value: clampLimit(limit)},
	}

	repoFilter := ""
	if len(repoIDs) > 0 {
		repoFilter = `
              AND repo_id IN (
                  SELECT id FROM repos
                  WHERE org_id = {org_id:String}
                    AND (repo IN {repo_ids:Array(String)} OR toString(id) IN {repo_ids:Array(String)})
              )`
		bindings = append(bindings, clickhouse.Binding{Name: "repo_ids", Value: repoIDs})
	}

	query := `
        SELECT
            reviewer,
            author,
            reviews_count,
            day,
            toString(repo_id) AS repo_id
        FROM (
            SELECT
                repo_id,
                reviewer,
                author,
                day,
                argMax(reviews_count, computed_at) AS reviews_count
            FROM review_edges_daily
            WHERE org_id = {org_id:String}
              AND day >= {since_date:Date}
              AND day <= {until_date:Date}` + repoFilter + `
            GROUP BY repo_id, reviewer, author, day
        )
        ORDER BY reviews_count DESC, repo_id, reviewer, author, day
        LIMIT {limit:UInt64}`

	rows, err := client.Query(ctx, query, bindings)
	if err != nil {
		return nil, fmt.Errorf("reviewedges: query: %w", err)
	}
	defer rows.Close()

	// Non-nil even with zero rows: the schema declares
	// edges: [ReviewEdgeRow!]! (non-null list) -- same "initialize
	// explicitly" convention featureflags.Resolve documents.
	edges := []model.ReviewEdgeRow{}
	for rows.Next() {
		var reviewer, author, repoID string
		// ClickHouse's `reviews_count` column is UInt32 (migration 004);
		// the native Go driver rejects scanning a UInt32 column into a
		// signed destination outright, the same class of mismatch
		// dev-health-go/readers' PullRequestStateRow.Number doc comment
		// documents. Scan into uint32, convert to int (the model field's
		// type) only after the value is safely in Go.
		var reviewsCount uint32
		var day time.Time
		if scanErr := rows.Scan(&reviewer, &author, &reviewsCount, &day, &repoID); scanErr != nil {
			return nil, fmt.Errorf("reviewedges: scan: %w", scanErr)
		}
		edge := model.ReviewEdgeRow{
			Reviewer:     reviewer,
			Author:       author,
			ReviewsCount: int(reviewsCount),
			Day:          graphqldate.New(day),
		}
		// Mirrors Python's `repo_id=str(row["repo_id"]) if
		// row.get("repo_id") else None`: only a non-empty string counts.
		// repo_id is a NOT NULL UUID column, so toString(repo_id) never
		// actually produces "" in practice, but the check is kept for
		// exact behavioral parity rather than assumed.
		if repoID != "" {
			id := repoID
			edge.RepoID = &id
		}
		edges = append(edges, edge)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("reviewedges: rows: %w", err)
	}

	return &model.ReviewEdgesResult{
		Edges:      edges,
		TotalCount: len(edges),
	}, nil
}
